File: app/src/main/java/com/arnaudpiroelle/manga/ui/history/HistoryPresenter.java
package com.arnaudpiroelle.manga.ui.history;

import com.arnaudpiroelle.manga.core.ui.presenter.Presenter;
import com.arnaudpiroelle.manga.model.History;

import java.util.Collections;
import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action0;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import se.emilsjolander.sprinkles.CursorList;
import se.emilsjolander.sprinkles.Model;
import se.emilsjolander.sprinkles.Query;

public class HistoryPresenter implements Presenter<History> {

    HistoryListingCallback callback;

    public HistoryPresenter(HistoryListingCallback callback) {
        this.callback = callback;
    }

    @Override
    public void list() {
        callback.onListingLoading();

        Observable.create(new Observable.OnSubscribe<List<History>>() {
            @Override
            public void call(Subscriber<? super List<History>> subscriber) {
                try (CursorList<History> historyCursorList = Query.all(History.class).get()) {


                    List<History> histories = historyCursorList.asList();

                    Collections.reverse(histories);
                    subscriber.onNext(histories);
                }

                subscriber.onCompleted();
            }
        })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(callback::onListingLoaded);
    }

    public void cleanHistory() {
        Observable.from(Query.all(History.class).get().asList())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        History::delete,
                        throwable -> {

                        },
                        this::list);
    }

    public interface HistoryListingCallback {
        void onListingLoading();

        void onListingLoaded(List<History> histories);
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/ui/history/HistoryPresenter.kt
package com.arnaudpiroelle.manga.ui.history

import com.arnaudpiroelle.manga.api.core.rx.plusAssign
import com.arnaudpiroelle.manga.core.db.dao.HistoryDao
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class HistoryPresenter(
        private val view: HistoryContract.View,
        private val historyDao: HistoryDao) : HistoryContract.UserActionsListener {

    private val disposable = CompositeDisposable()

    override fun register() {
        disposable += historyDao.getAll()
                .subscribeOn(Schedulers.io())
                .map { it.reversed() }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(view::displayHistories, {})
    }

    override fun unregister() {
        disposable.clear()
    }

    override fun cleanHistory() {
        disposable += Completable.fromAction { historyDao.deleteAll() }
                .subscribeOn(Schedulers.io())
                .subscribe({}, {})
    }
}
